### DISCUSSION TITLE
Mkdocs github pages with travis - how can I use mike commands in travis?

### DISCUSSION THREAD

**Discussion by rajdevworks** [jimporter/mike#198] [Q&A]
I have my github pages repo with mkdocs and using travis for CI.
When I make an update to the current branch on UI, it stays in the repo, but is not rendered on the published site.
However, when I make a change in the local branch, run mike commands (deploy, set-default), and then push it to the repo, it works.

I tried adding the mike commands after mkdocs build and it failed with 1. What am I doing wrong?

Edit. I made progress and looks like Permission issue for mike to write back to gh-pages branch.
```
$ mkdocs build -f docs/mkdocs-current.yml
$ mike deploy -F docs/mkdocs-current.yml --push --update-aliases current latest
INFO    -  Cleaning site directory
INFO    -  Building documentation to directory: /home/travis/build/<repo-name>/test-mkdocs-versioniong/docs/site
INFO    -  Documentation built in 2.62 seconds
error: failed to push branch gh-pages to origin:
  remote: Write access to repository not granted.
  fatal: unable to access 'https://github.com/<repo-name>/test-mkdocs-versioniong.git/': The requested URL returned error: 403
The command "mike deploy -F docs/mkdocs-current.yml --push --update-aliases current latest" failed and exited with 1 during .
```

**Comment by jimporter** [comment]
First, you need to use `mike` to build your docs. See [the README section](https://github.com/jimporter/mike#building-your-docs). You may also want to fetch your `gh-pages` branch (see [this other README section](https://github.com/jimporter/mike#deploying-via-ci)).

**Reply by rajdevworks** [reply]
I have these in `before_deploy` section of travis.yml
```
before_deploy:
  - git config user.name '<user>'
  - git config user.email '<user-email>'
  - git fetch origin gh-pages --depth=1

  - mike deploy 0.2 current -F docs/mkdocs.yml
  - mike deploy 0.1 stable -F docs/mkdocs-0.1.0.yml
  - mike set-default -F docs/mkdocs.yml --push current
  - mike deploy -F docs/mkdocs-0.1.0.yml  --push --update-aliases 0.1 stable
 ```

On my travis build, I see these errors:

```$ git config user.name '<user>'
$ git config user.email '<user-email>'
$ git fetch origin gh-pages --depth=1
$ mike deploy 0.2 current -F docs/mkdocs.yml
$ mike deploy 0.1 stable -F docs/mkdocs-0.1.0.yml
$ mike set-default -F docs/mkdocs.yml --push current
error: failed to push branch gh-pages to origin:
  To github.com:myorg/test-mkdocs-versioniong.git
   ! [rejected]        gh-pages -> gh-pages (non-fast-forward)
  error: failed to push some refs to 'github.com:myorg/test-mkdocs-versioniong.git'
  hint: Updates were rejected because a pushed branch tip is behind its remote
  hint: counterpart. If you want to integrate the remote changes, use 'git pull'
  hint: before pushing again.
  hint: See the 'Note about fast-forwards' in 'git push --help' for details.
The command "mike set-default -F docs/mkdocs.yml --push current" failed and exited with 1 during .```

It works perfectly on my local workstation, but it fails on travis build.

**Reply by jimporter** [reply]
Unfortunately, I don't know much of anything about Travis and can't help you with the specifics there. (I used to use it, but that was a long time ago, before private equity decided to bleed it dry.)

Here are some more general debugging tips though. Make sure your local clone of `gh-pages` is actually up to date at that point, e.g. by adding `git log -4` to print the recent commits, and comparing the output with what you see in your Git forge. I'm also a bit confused about why you're running `mike deploy` for the 0.1 version twice. There shouldn't be any need to do that.
